frame/vesting/src/lib.rs

```diff
@@ -300,72 +405,206 @@ impl<T: Config> VestedTransfer for Pallet<T> {
 	type Moment = MomentOf<T>;
 	type Balance = BalanceOf<T>;
 	type MinVestedTransfer = T::MinVestedTransfer;
+	type VestingScheduleId = T::VestingScheduleId;
+	type VestingScheduleNonce = VestingScheduleNonce<T>;
 
 	#[transactional]
 	fn vested_transfer(
 		asset: Self::AssetId,
 		from: &Self::AccountId,
 		to: &Self::AccountId,
-		schedule: VestingSchedule<Self::BlockNumber, Self::Moment, Self::Balance>,
+		schedule_input: VestingScheduleInput<Self::BlockNumber, Self::Moment, Self::Balance>,
 	) -> frame_support::dispatch::DispatchResult {
 		ensure!(from != to, Error::<T>::TryingToSelfVest);
 
+		let vesting_schedule_id = Self::VestingScheduleNonce::increment()?;
+		let schedule = VestingSchedule::from_input(vesting_schedule_id, schedule_input);
+
 		let schedule_amount = ensure_valid_vesting_schedule::<T>(&schedule)?;
 
-		let total_amount = Self::locked_balance(to, asset)
-			.checked_add(&schedule_amount)
-			.ok_or(ArithmeticError::Overflow)?;
+		let locked = Self::locked_balance(to, asset, VestingScheduleIdSet::All)
+			.unwrap_or_else(|_| Zero::zero());
+
+		let total_amount = locked.safe_add(&schedule_amount)?;
 
 		T::Currency::transfer(asset, from, to, schedule_amount)?;
 		T::Currency::set_lock(VESTING_LOCK_ID, asset, to, total_amount)?;
-		<VestingSchedules<T>>::try_append(to, asset, &schedule)
-			.map_err(|_| Error::<T>::MaxVestingSchedulesExceeded)?;
+
+		<VestingSchedules<T>>::mutate(to, asset, |schedules| {
+			schedules
+				.try_insert(vesting_schedule_id, schedule.clone())
+				.map_err(|_| Error::<T>::MaxVestingSchedulesExceeded)
+		})?;
 
 		Self::deposit_event(Event::VestingScheduleAdded {
 			from: from.clone(),
 			to: to.clone(),
 			asset,
 			schedule,
+			vesting_schedule_id,
 		});
 
 		Ok(())
 	}
 }
 
 impl<T: Config> Pallet<T> {
-	fn do_claim(who: &AccountIdOf<T>, asset: AssetIdOf<T>) -> Result<BalanceOf<T>, DispatchError> {
-		let locked = Self::locked_balance(who, asset);
-		if locked.is_zero() {
+	fn do_claim(
+		who: &AccountIdOf<T>,
+		asset: AssetIdOf<T>,
```

Comment:
not sure if this API is good. you have to tell what assets you what to claim and what schedules. while claim all assists and claim all schedules is best thing. i may be claim with to parameters so it can be bot automated to some aggregated account.